qubes/device_protocol.py

```diff
@@ -0,0 +1,948 @@
+
+    @property
+    def description(self) -> str:
+        """
+        Short human-readable description.
+
+        For unknown device returns `unknown device (unknown vendor)`.
+        For unknown USB device returns `unknown usb device (unknown vendor)`.
+        For unknown USB device with known serial number returns
+            `<serial> (unknown vendor)`.
+        """
+        if self.product and self.product != "unknown":
+            prod = self.product
+        elif self.name and self.name != "unknown":
+            prod = self.name
+        elif self.serial and self.serial != "unknown":
+            prod = self.serial
+        elif self.parent_device is not None:
+            return f"sub-device of {self.parent_device}"
+        else:
+            prod = f"unknown {self.devclass if self.devclass else ''} device"
+
+        if self.vendor and self.vendor != "unknown":
+            vendor = self.vendor
+        elif self.manufacturer and self.manufacturer != "unknown":
+            vendor = self.manufacturer
+        else:
+            vendor = "unknown vendor"
+
+        return f"{prod} ({vendor})"
+
+    @property
+    def interfaces(self) -> List[DeviceInterface]:
+        """
+        Non-empty list of device interfaces.
+
+        Every device should have at least one interface.
+        """
+        if not self._interfaces:
+            return [DeviceInterface.unknown()]
+        return self._interfaces
+
+    @property
+    def parent_device(self) -> Optional[Device]:
+        """
+        The parent device if any.
+
+        If the device is part of another device (e.g. it's a single
+        partition of an usb stick), the parent device id should be here.
+        """
+        return self._parent
+
+    @property
+    def subdevices(self) -> List['DeviceInfo']:
+        """
+        The list of children devices if any.
+
+        If the device has subdevices (e.g. partitions of an usb stick),
+        the subdevices id should be here.
+        """
+        return [dev for dev in self.backend_domain.devices[self.devclass]
+                if dev.parent_device.ident == self.ident]
+
+    @property
+    def attachment(self) -> Optional[QubesVM]:
+        """
+        VM to which device is attached (frontend domain).
+        """
+        return self._attachment
+
+    def serialize(self) -> bytes:
+        """
+        Serialize an object to be transmitted via Qubes API.
+        """
+        # 'backend_domain', 'attachment', 'interfaces', 'data', 'parent_device'
+        # are not string, so they need special treatment
+        default_attrs = {
+            'ident', 'devclass', 'vendor', 'product', 'manufacturer', 'name',
+            'serial', 'self_identity'}
+        properties = b''
+        for key, value in ((key, getattr(self, key)) for key in default_attrs):
+            properties = self.pack_property(properties, key, value)
+        # remove unnecessary space in the beginning
+        properties = properties[1:]
```

Comment:
This is a bit weird - `self.pack_property` gets already serialized `properties`, yet still adds the leading space. And in fact, it's a bit weird to have that parameter, since it isn't used for anything besides appending result to it.
Isn't it better to `self.pack_property` really pack a single property (without concatenating to earlier content) and then do `b' '.join()` on results? Which would be also faster (not sure if that really matters here).